### PR TITLE
Signalling extensions from SQF

### PR DESCRIPTION
**Pull request by esteldunedain**
This PR adds a way of communicating data from SQF to an extension through Intercept.

I'm not sure this is the correct/best way to do it, so please review. 

~~Besides, it's crashing like a test dummy; but apparently most Intercept is doing that today. I'll try to pinpoint the cause.~~

### CODE REVIEW

**Comment by Glowbal** [comment]
What would be a use case in where this would be benefitial to have?

**Comment by esteldunedain** [comment]
The use case I was thinking about is to better integrate extensions with existing sqf frameworks: e.g. enabling/disabling features based on the ACE_settings framework, passing custom events, etc.

Alternatively, those things could be done asynchronously through GVARS, but that requires polling the variable all frames from the extension to notice changes.

**Comment by Glowbal** [comment]
Thanks @esteldunedain. I think it's a good things to support this in intercept. It ideally wouldn't be necessary but I can see it offer some flexibility.

@NouberNou will probably take a look at this soon and determine if this is something we would like in intercept.

File: src/host/invoker/invoker.cpp [hunk 1]
```diff
@@ -142,6 +143,25 @@ namespace intercept {
...
+
+        auto modules = extensions::get().modules();
+        if (modules.find(extension_name) == modules.end()) {
+            return false;
+        }
+        auto module = modules[extension_name];
+
+        _signal_params = invoke_raw("getvariable", &_mission_namespace, "NAMESPACE", &game_value("intercept_signal_var"), "STRING");
+        if (module.functions.on_signal) {
+            module.functions.on_signal(_signal_params);
```
Comment (Glowbal) [inline src/host/invoker/invoker.cpp]: May need to check if `_signal_params` is a valid value?

File: rv/addons/core/config.cpp [hunk 2]
```diff
@@ -37,5 +37,13 @@ class CfgFunctions
                 headerType = -1;
             };
         };
+        class api
+        {
+            class signal
+            {
+                file = "z\intercept\rv\addons\core\signal.sqf";
```
Comment (overfl0) [inline rv/addons/core/config.cpp]: Shouldn't there be double backslashes?

Comment (Itsyuka) [inline rv/addons/core/config.cpp]: @overfl0 You'd think, but the cpp/hpp/h files for Arma 3 doesn't use escapes as we traditionally see them as.

File: src/client/intercept/intercept.hpp [hunk 3]
```diff
@@ -15,6 +15,7 @@ namespace intercept {
         DLLEXPORT void __cdecl mission_end();
         DLLEXPORT void __cdecl mission_stopped();
         DLLEXPORT void __cdecl on_frame();
+        DLLEXPORT void __cdecl on_signal(std::string &signal_name_, game_value& value1_);
```
Comment (esteldunedain) [inline src/client/intercept/intercept.hpp]: I recon we don't want `std::string`s over the ABI, but this string isn't a `rv_string` on the first place and I don't know if we want to a dd yet another string type to Intercept. Maybe add a contructor to make `rv_string`from `std::string`?

Comment (esteldunedain) [inline src/client/intercept/intercept.hpp]: Mmm, would this work?:
```c++
rv_string rv_signal_name(game_value(signal_name));
```

File: src/host/invoker/invoker.cpp [hunk 4]
```diff
@@ -142,6 +143,26 @@ namespace intercept {
...
+        std::string signal_name = args_.as_string(1);
+        LOG(DEBUG) << "Signal " << extension_name << " : " << signal_name << " START";
+
+        auto modules = extensions::get().modules();
+        if (modules.find(extension_name) == modules.end()) {
+            return false;
+        }
+        auto module = modules[extension_name];
+        _invoker_unlock signal_lock(this);
+        _signal_params = invoke_raw("getvariable", &_mission_namespace, "NAMESPACE", &game_value("intercept_signal_var"), "STRING");
```
Comment (esteldunedain) [inline src/host/invoker/invoker.cpp]: I'm still trying to figure out why this line 158 makes Arma CTD. If instead I replace it with:
```c++_signal_params = game_value(std::string("test"));

**Comment by NouberNou** [comment]
Is this working now that the crash bug was fixed?

Also, what would make this even more useful is if you could specify the name of the exported function. I'd suggest using GetProcAddress to return the address of the function the first time and then storing the address in a unordered_map (since looking up in that is going to be faster than GetProcAddress) for subsequent calls.

Also I'd like to investigate how to return variables from this.

**Comment by NouberNou** [comment]
Gunna merge this in and work on adding return variables and user defined function names.

**Pull request merged**